Issue/45/inform option

sschmidt23 opened the pull request:
This simple PR addresses one isolated portion of issue #45 (the main one, but discussion in that issue grew beyond the specific topic).  It adds a new `inform_options` dictionary to the yaml config file that control whether to write out a trained model or read in a trained model if a pre-trained model exists.  This PR does *not* address the provenance concerns that come with creating/ingesting pre-trained models, that will be left for a future update following discussions of what provenance tracking system that we want to use.

I added tests that create a model and then compare the outputs when ingesting that model vs an initial run, so it looks like everything functions as it should.

johannct commented on tests/test_algos.py:
```diff
@@ -9,6 +10,10 @@
+    Run inform, write temporary trained model to
+    'tempmodelfile.tmp', run photo-z algorithm.
+    Then, load temp modelfile and re-run, return
+    both datasets.
```
I wonder if it would not be better to have a dedicated test for loading, for saving, etc.... I understand it is painful, if you want I can implement it to finsih the review. I am afraid that quickly the test calls become too complex to easily point to regression issues in the future.

sschmidt23 replied:
Separate tests for loading and saving would be good, this test was more for actually checking that the results from loading a trained model agree with the results from computing the model.  Were you thinking of e.g tests where self.modelfile is not defined and such?

johannct replied:
I was thinking of all the cases where codecov notices lack of coverage : try/except blocks on loading and saving pretrained estimators

sschmidt23 replied:
I just added tests that I think cover all the examples of saving and loading (the one_algo function saves a model, then loads the model and compares the two outputs, so save/load is tested in the course of that), plus tests checking the try/except for missing keys and files that weren't covered before.  So, I think we're now good.

Can you take a quick look and see if there are any cases that I've missed?  codecov is happy, but I'm trying to think if there might be an additional missing use case.

johannct replied:
LGTM thanks Sam. Remember to modify the action as well (to push on master)

johannct commented on rail/estimation/algos/flexzboost.py:
```diff
@@ -141,6 +143,14 @@ def inform(self):
                 bestsharp = sharp
         model.sharpen_alpha = bestsharp
         self.model = model
+        if self.inform_options['save_train']:
```
we need some boilerplate here, unfortunately, to check early (before starting to train) that modelfile is provided if save_train or load_train is True. I just experienced wasting training time by ending with:
```
finding best sharpen parameter...
Traceback (most recent call last):
  File "main.py", line 67, in <module>
    main(sys.argv)
  File "main.py", line 43, in main
    pz.inform()
  File "/home/cohen/lsst/codes/pipelines/photoz/RAIL/rail/estimation/algos/flexzboost.py", line 147, in inform
    with open(self.inform_options['modelfile'], 'wb') as f:
KeyError: 'modelfile'
```

sschmidt23 replied:
I added a check in the __init__ for the two codes that write out a model, and similarly put a check on the modelfile in estimator.py in load_pretrained_model to check that the `modelfile` keyword is there and that the file exists.  These are not currently covered by the tests, which is related to your comment above about separate tests, I think.

sschmidt23 replied:
We can add tests of coverage once you confirm that this answers your initial comment

sschmidt23 replied:
coverage tests were added so I'll resolve this

johannct commented on rail/estimation/algos/flexzboost.py:
```diff
@@ -141,6 +143,14 @@ def inform(self):
+                pickle.dump(file=f, obj=model,
+                            protocol=pickle.HIGHEST_PROTOCOL)
+
+    def load_pretrained_model(self):
```
what code is in charge of calling `load_pretrained_model`, and based on which set of parameters? In the tests we probably need to exercise : 1/ train and stop (no estimate); 2/ train and estimate; 3/ load and estimate;

sschmidt23 replied:
I was reworking this a bit in separate branch, but load_pretrained_model is handled in main.py currently, and yes, separate tests are probably necessary.  I wasn't thinking of "train and stop" as an option, but I guess that is a potential use case.

johannct reviewed (changes requested):
thanks for this Sam, let's try to have a quick turnaround on my comments. I am available to pair code if we find a good time (though next week is going to be tough)

sschmidt23 commented on examples/main.py:
```diff
@@ -34,7 +34,14 @@ def main(argv):
+        # note: specific options set in subclasss func def
+        pz.load_pretrained_model()
+    else:
+        pz.inform()
```
I think a jupyter notebook that will be added with issue #51 will take care of this, having examples written out in a notebook I think will suffice, but I'll say lets leave it for that separate issue.

johannct replied:
ok can you put a note there?

sschmidt23 commented on .github/workflows/main.yml:
```diff
@@ -5,7 +5,7 @@ name: rail
 on:
   push:
-    branches: [ master ]
+    branches: [ issue/45/inform_option ]
```
will fix this once the PR is accepted

johannct replied:
sure

sschmidt23 commented on rail/estimation/estimator.py:
```diff
@@ -64,6 +64,13 @@ def inform(self):
+        load the information required to run estimate
+        """
+        raise NotImplementedError  # pragma: no cover
+
```
I moved the 'boilerplate' to estimator.py and put a note that subclasses can overwrite if they are loading something other than a pickle file

johannct reviewed (approved):
LGTM, just these items :

sschmidt23 commented:
I added a note on issue #51 and added some new tests.  @johannct can you take one quick look at the tests to see if I've covered everything from a user point of view, I think I have but a quick eye on the code to be sure is good.  Once you sign off I'll resolve the merge conflict and merge.

eacharles reviewed (approved):
Thanks.  Maybe:
1. Point out that ceci gets installed along with rail_base, so you can do pip install rail_base and that will get you both
2. Clarify what “point at the root of rail” means.   I think that is just the same as activativectge conda env, but I’m not sure
